src/plugins/intel_cpu/src/nodes/gather.cpp

```diff
@@ -789,46 +717,68 @@ void Gather::execWeightsCompressed8bit() {
     });
 }
 
-void Gather::execWeightsCompressed() {
+int8_t Gather::get_i4(const uint8_t& val, bool high) {
+    if (high) {
+        if (val & 0x80) {
+            return static_cast<int8_t>(((val >> 4) & 0x7) | 0xf8);
+        } else {
+            return static_cast<int8_t>((val & 0xF) >> 4);
+        }
+    }
+    if (val & 0x8) {
+        // Just fill in the high 4 bits with 1
+        return static_cast<int8_t>((val & 0x7) | 0xf8);
+    } else {
+        return static_cast<int8_t>(val & 0xF);
+    }
+}
+int8_t Gather::get_u4(const uint8_t& val, bool high) {
+    if (high) {
+        return (val >> 4) & 0xF;
+    }
+    return val & 0xF;
+}
+
+void Gather::execCompressed() {
     auto in_precison = getParentEdgeAt(GATHER_DATA)->getMemoryPtr()->getPrecision();
     auto out_precision = getChildEdgeAt(0)->getMemoryPtr()->getPrecision();
 
     if (out_precision == ov::element::f16) {
```

Comment:
Since we are now instancing several implementations we need to apply conditional compilation optimization to prevent binary size increase. Please see https://github.com/openvinotoolkit/openvino/pull/23287/files#diff-37239f1a7296b799b8e22fdf68077d40bcb78bc48dbbad2bd57ef2f76fca905aR377 fo reference

Reply:
Done.